regressions with llcreader

I am getting some regressions with the latest release of fsspec on my xmitgcm package.
https://travis-ci.org/xgcm/xmitgcm/jobs/572386378


I haven't had time to look into this is too much depth, but the errors look like

```
E           KeyError: 'https://storage.googleapis.com/pangeo-ecco/llc/masks/llc_2160_masks.zarr/.zmetadata'
...
E           TypeError: request() got an unexpected keyword argument 'size_policy'
```

I thought I would report this here while I dig further.

--- a/fsspec/implementations/http.py
+++ b/fsspec/implementations/http.py
@@ -24,7 +24,7 @@ class HTTPFileSystem(AbstractFileSystem):
     sep = '/'
 
     def __init__(self, simple_links=True, block_size=None, same_scheme=True,
-                 **storage_options):
+                 size_policy=None, **storage_options):
         """
         Parameters
         ----------
@@ -37,6 +37,7 @@ class HTTPFileSystem(AbstractFileSystem):
         same_scheme: True
             When doing ls/glob, if this is True, only consider paths that have
             http/https matching the input URLs.
+        size_policy: this argument is deprecated
         storage_options: key-value
             May be credentials, e.g., `{'auth': ('username', 'pword')}` or any
             other parameters passed on to requests

--- a/fsspec/implementations/local.py
+++ b/fsspec/implementations/local.py
@@ -162,19 +162,42 @@ def make_path_posix(path, sep=os.sep):
 class LocalFileOpener(object):
     def __init__(self, path, mode, autocommit=True, fs=None, **kwargs):
         self.path = path
+        self.mode = mode
         self.fs = fs
         self.autocommit = autocommit
-        if autocommit or 'w' not in mode:
-            self.f = open(path, mode=mode)
+        self._open()
+
+    def _open(self):
+        if self.autocommit or 'w' not in self.mode:
+            self.f = open(self.path, mode=self.mode)
         else:
             # TODO: check if path is writable?
             i, name = tempfile.mkstemp()
             self.temp = name
-            self.f = open(name, mode=mode)
-        if 'w' not in mode:
-            self.details = self.fs.info(path)
+            self.f = open(name, mode=self.mode)
+        if 'w' not in self.mode:
+            self.details = self.fs.info(self.path)
             self.size = self.details['size']
             self.f.size = self.size
+
+    def __setstate__(self, state):
+        if 'r' in state['mode']:
+            loc = self.state.pop('loc')
+            self._open()
+            self.f.seek(loc)
+        else:
+            self.f = None
+        self.__dict__.update(state)
+
+    def __getstate__(self):
+        d = self.__dict__.copy()
+        d.pop('f')
+        if 'r' in self.mode:
+            d['loc'] = self.f.tell()
+        else:
+            if not self.f.closed:
+                raise ValueError("Cannot serialise open write-mode local file")
+        return d
 
     def commit(self):
         if self.autocommit:

--- a/fsspec/implementations/tests/test_http.py
+++ b/fsspec/implementations/tests/test_http.py
@@ -83,6 +83,12 @@ def test_list(server):
     assert out == [server + '/index/realfile']
 
 
+def test_policy_arg(server):
+    h = fsspec.filesystem('http', size_policy='get')
+    out = h.glob(server + '/index/*')
+    assert out == [server + '/index/realfile']
+
+
 def test_exists(server):
     h = fsspec.filesystem('http')
     assert not h.exists(server + '/notafile')

--- a/fsspec/spec.py
+++ b/fsspec/spec.py
@@ -2,6 +2,8 @@ from hashlib import md5
 import io
 import os
 import logging
+
+from .transaction import Transaction
 from .utils import read_block, tokenize, stringify_path
 logger = logging.getLogger('fsspec')
 
@@ -787,47 +789,6 @@ class AbstractFileSystem(up):
     def clear_instance_cache(cls):
         """Remove cached instances from the class cache"""
         cls._cache.clear()
-
-
-class Transaction(object):
-    """Filesystem transaction write context
-
-    Gathers files for deferred commit or discard, so that several write
-    operations can be finalized semi-atomically. This works by having this
-    instance as the ``.transaction`` attribute of the given filesystem
-    """
-
-    def __init__(self, fs):
-        """
-        Parameters
-        ----------
-        fs: FileSystem instance
-        """
-        self.fs = fs
-        self.files = []
-
-    def __enter__(self):
-        self.start()
-
-    def __exit__(self, exc_type, exc_val, exc_tb):
-        """End transaction and commit, if exit is not due to exception"""
-        # only commit if there was no exception
-        self.complete(commit=exc_type is None)
-        self.fs._intrans = False
-
-    def start(self):
-        """Start a transaction on this FileSystem"""
-        self.fs._intrans = True
-
-    def complete(self, commit=True):
-        """Finish transaction: commit or discard all deferred files"""
-        for f in self.files:
-            if commit:
-                f.commit()
-            else:
-                f.discard()
-        self.files = []
-        self.fs._intrans = False
 
 
 class AbstractBufferedFile(io.IOBase):

--- a/fsspec/transaction.py
+++ b/fsspec/transaction.py
@@ -1,0 +1,79 @@
+class Transaction(object):
+    """Filesystem transaction write context
+
+    Gathers files for deferred commit or discard, so that several write
+    operations can be finalized semi-atomically. This works by having this
+    instance as the ``.transaction`` attribute of the given filesystem
+    """
+
+    def __init__(self, fs):
+        """
+        Parameters
+        ----------
+        fs: FileSystem instance
+        """
+        self.fs = fs
+        self.files = []
+
+    def __enter__(self):
+        self.start()
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        """End transaction and commit, if exit is not due to exception"""
+        # only commit if there was no exception
+        self.complete(commit=exc_type is None)
+        self.fs._intrans = False
+
+    def start(self):
+        """Start a transaction on this FileSystem"""
+        self.fs._intrans = True
+
+    def complete(self, commit=True):
+        """Finish transaction: commit or discard all deferred files"""
+        for f in self.files:
+            if commit:
+                f.commit()
+            else:
+                f.discard()
+        self.files = []
+        self.fs._intrans = False
+
+
+class FileActor(object):
+
+    def __init__(self):
+        self.files = []
+
+    def commit(self):
+        for f in self.files:
+            f.commit()
+        self.files.clear()
+
+    def discard(self):
+        for f in self.files:
+            f.discard()
+        self.files.clear()
+
+    def append(self, f):
+        self.files.append(f)
+
+
+class DaskTransaction(Transaction):
+    def __init__(self, fs):
+        """
+        Parameters
+        ----------
+        fs: FileSystem instance
+        """
+        import distributed
+        super().__init__(fs)
+        client = distributed.default_client()
+        self.files = client.submit(FileActor, actor=True).result()
+
+    def complete(self, commit=True):
+        """Finish transaction: commit or discard all deferred files"""
+        if commit:
+            self.files.commit().result()
+        else:
+            self.files.discard().result()
+        self.fs._intrans = False

